src/sorunlib/wiregrid.py

```diff
@@ -198,6 +198,112 @@ def _check_temperature_sensors():
     _verify_temp_response(resp, 'AIN2C', 0)
 
 
+def _check_wiregrid_position():
+    """Check the wiregrid position."""
+    actuator = run.CLIENTS['wiregrid']['actuator']
+    resp = actuator.acq.status()
+    ls_status = resp.session['data']['fields']['limitswitch']
+    if (ls_status['LSR1'] == 1 or ls_status['LSL1'] == 1) and not (ls_status['LSR2'] == 1 and ls_status['LSL2'] == 1):
+        position = 'outside'
+    elif (ls_status['LSR2'] == 1 or ls_status['LSL2'] == 1) and not (ls_status['LSR1'] == 1 or ls_status['LSL1'] == 1):
+        position = 'inside'
+    else:
+        raise RuntimeError("The wiregrid position is unknown. The limitswitch response is like this:\n" +
+                           "LSR1: {}\n".format('ON' if ls_status['LSR1'] == 1 else 'OFF')
+                           + "LSL1: {}\n".format('ON' if ls_status['LSL1'] == 1 else 'OFF')
+                           + "LSR2: {}\n".format('ON' if ls_status['LSR2'] == 1 else 'OFF')
+                           + "LSL2: {}\n".format('ON' if ls_status['LSL2'] == 1 else 'OFF')
+                           + "Aborting...")
+    return position
+
+
+def _stop_hwp_with_wiregrid():
+    """Stop the HWP after comfirming the wiregrid is inserted."""
+    # Check the wiregrid position
+    position = _check_wiregrid_position()
+    if position == 'outside':
+        raise RuntimeError("The wiregrid is not inserted. Aborting...")
+    elif position == 'inside':
+        # Stop the HWP
+        print("Starting to stop the HWP.")
+        run.hwp.stop(active=True)
+        print("The HWP stopped successfully.")
+    else:
+        raise RuntimeError("Unknown wiregrid position. Aborting...")
+
+
+def _spin_hwp_with_wiregrid(target_hwp_direction):
+    """Spin the HWP to the target direction at 2Hz after comfirming the wiregrid is inserted.
+
+    Args:
+        target_hwp_direction (str): Target HWP direction, 'forward' or 'backward'.
+    """
+    # Check the wiregrid position
+    position = _check_wiregrid_position()
```

Comment:
The checks in both `_stop_hwp_with_wiregrid()` and `_spin_hwp_with_wiregrid()` seem like are duplicating the check, and as a result making the functions more complicated than they need to be. They're only used in `_reverse_hwp_with_wiregrid()`, which is only called after `insert()`. If `insert()` fails to insert the wiregrid it raises and the schedule with exit.

If you still want to check the wiregrid is inserted, I'd say do the check in `_reverse_hwp_with_wiregrid()` before doing anything else. You can then assume its inserted in these other two functions, and they simplify down to:
```python
def _stop_hwp():
    """Stop the HWP."""
    print("Starting to stop the HWP.")
    run.hwp.stop(active=True)
    print("The HWP stopped successfully.")

```

```python
def _spin_hwp(target_hwp_direction):
    """Spin the HWP to the target direction at 2Hz.
    Args:
        target_hwp_direction (str): Target HWP direction, 'forward' or 'backward'.
    """
    if target_hwp_direction == 'forward':
        print("Starting to spin up the HWP in forward.")
        run.hwp.set_freq(freq=2.0)
        print("The HWP is spinning at 2Hz in forward successfully.")
    elif target_hwp_direction == 'backward':
        print("Starting to spin up the HWP in backward.")
        run.hwp.set_freq(freq=-2.0)
        print("The HWP is spinning at 2Hz in backward successfully.")
```